### PR TITLE
8291457: Clean up lifecycle of CDS dump-time data structures

### PR DESCRIPTION
**Pull request by iklam**
I've rewritten the following data CDS structures to follow the [C++ rule-of-three](https://en.cppreference.com/w/cpp/language/rule_of_three) to improve their lifecycle handling and fix a memory leak. This is done to prepare for future CDS changes that make more allocations during archive creation.

> Rule of Three: If a class requires a user-defined destructor, a user-defined copy constructor, or a user-defined copy assignment operator, it almost certainly requires all three.

- DumpTimeClassInfo
- DumpTimeClassInfo::DTLoaderConstraint
- DumpTimeClassInfo::DTVerifierConstraint
- DumpTimeLambdaProxyClassInfo

I implemented some of the assignment operators using the [copy-and-swap idiom](https://en.cppreference.com/w/cpp/language/copy_assignment) to handle `Symbol` refcounting. Thanks to @kimbarrett for the suggestion.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291457](https://bugs.openjdk.org/browse/JDK-8291457): Clean up lifecycle of CDS dump-time data structures

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [29661dd9](https://git.openjdk.org/jdk/pull/9722/files/29661dd97f0a63bc9b29ed09d7af291d59825b65)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [29661dd9](https://git.openjdk.org/jdk/pull/9722/files/29661dd97f0a63bc9b29ed09d7af291d59825b65)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9722/head:pull/9722` \
`$ git checkout pull/9722`

Update a local copy of the PR: \
`$ git checkout pull/9722` \
`$ git pull https://git.openjdk.org/jdk pull/9722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9722`

View PR using the GUI difftool: \
`$ git pr show -t 9722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9722.diff">https://git.openjdk.org/jdk/pull/9722.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/dumpTimeClassInfo.hpp [hunk 1]
```diff
@@ -40,32 +40,93 @@ class DumpTimeClassInfo: public CHeapObj<mtClass> {
...
+
+  static void maybe_increment_refcount(Symbol* s) {
+    if (s != NULL) {
+      s->increment_refcount();
+    }
+  }
+  static void maybe_decrement_refcount(Symbol* s) {
+    if (s != NULL) {
+      s->decrement_refcount();
+    }
```
Comment (coleenp) [inline src/hotspot/share/cds/dumpTimeClassInfo.hpp:52]: Can you put these two functions as inline functions in symbol.hpp.  We do this way too much in lots of code.

File: src/hotspot/share/cds/dumpTimeClassInfo.cpp [hunk 2]
```diff
@@ -31,40 +31,58 @@
...
-    for (int i = 0; i < clone_num_loader_constraints; i++) {
-      clone._loader_constraints->append(_loader_constraints->at(i));
+
+  {
+    int n = src.num_loader_constraints();
+    if (n > 0) {
+      _loader_constraints = new (ResourceObj::C_HEAP, mtClass) GrowableArray<DTLoaderConstraint>(n, mtClass);
+      for (int i = 0; i < n; i++) {
+        _loader_constraints->append(src._loader_constraints->at(i));
+      }
```
Comment (coleenp) [inline src/hotspot/share/cds/dumpTimeClassInfo.cpp:72]: Do you need the src copy after the copy constructor runs?  You could transfer ownership of the verifier and loader_constraints arrays if not.

File: src/hotspot/share/cds/dumpTimeClassInfo.cpp [hunk 3]
```diff
@@ -31,40 +31,58 @@
...
+
+DumpTimeClassInfo::~DumpTimeClassInfo() {
+  if (_verifier_constraints != NULL) {
+    assert(_verifier_constraint_flags != NULL, "must be");
+    delete _verifier_constraints;
+    delete _verifier_constraint_flags;
+  }
+  if (_loader_constraints != NULL) {
+    delete _loader_constraints;
+  }
```
Comment (coleenp) [inline src/hotspot/share/cds/dumpTimeClassInfo.cpp:85]: So this delete operator in this destructor calls the destructors for all the elements of the _verifier_constraints, _verifier_constraint_flags and _loader_constraints elements, which decrements refcounts in them, right?

**Review by coleenp — COMMENTED** [review]
This looks a lot cleaner. I have a couple of questions.

File: src/hotspot/share/cds/dumpTimeClassInfo.hpp [hunk 4]
```diff
@@ -40,32 +40,93 @@ class DumpTimeClassInfo: public CHeapObj<mtClass> {
...
+      maybe_increment_refcount(_name);
     }
-    DTLoaderConstraint() : _name(NULL), _loader_type1('0'), _loader_type2('0') {}
+    DTLoaderConstraint(const DTLoaderConstraint& src) {
+      _name = src._name;
+      _loader_type1 = src._loader_type1;
+      _loader_type2 = src._loader_type2;
+      maybe_increment_refcount(_name);
+    }
+    DTLoaderConstraint& operator=(DTLoaderConstraint src) {
```
Comment (calvinccheung) [inline src/hotspot/share/cds/dumpTimeClassInfo.hpp:69]: Could the src be passed as reference?

Comment (iklam) [reply src/hotspot/share/cds/dumpTimeClassInfo.hpp:69]: To use the copy-and-swap idiom, which makes the assignment operator reliably, src needs to be passed by value. See https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom

File: src/hotspot/share/cds/dumpTimeClassInfo.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/dumpTimeClassInfo.cpp:74]: The src copy is still needed after the copy. I added more comments to explain why the deep copy is necessary.

File: src/hotspot/share/cds/dumpTimeClassInfo.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/cds/dumpTimeClassInfo.cpp:87]: Yes. `_loader_constraints` is a GrowableArray, which calls its element type's destructor.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 5]
```diff
@@ -1472,11 +1473,24 @@ class CloneDumpTimeLambdaProxyClassTable: StackObj {
...
+// pointers are changed to use "buffer" addresses.)
+//
+// We save a copy of these tables and restore them after the dumping is finished.
+// This makes it possible to repeat the dumping operation (e.g., use
+// "jcmd VM.cds dynamic_dump" multiple times on the same JVM process).
+//
+// We use the copy constructors to clone the values in these tables. The copy constructors
+// must make a deep copy, as internal data structures such as the contents of
+// DumpTimeClassInfo::_loader_constraints are also modified by the ArchiveBuilder.
+
```
Comment (coleenp) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1493]: Ok, I see.  So it can't make a shallow copy.

**Review by coleenp — APPROVED** [review]
Looks good. Thanks for the comment.

**Review by calvinccheung — APPROVED** [review]
Looks good.

**Comment by iklam** [comment]
Thanks to @coleenp and @calvinccheung for the review.
/integrate

**Pull request closed without merging**